lib/min_heap.rb

```diff
@@ -1,73 +1,104 @@
 class HeapNode
   attr_reader :key, :value
-
+  
   def initialize(key, value)
     @key = key
     @value = value
   end
 end
 
 class MinHeap
-
+  
   def initialize
     @store = []
   end
-
+  
   # This method adds a HeapNode instance to the heap
-  # Time Complexity: ?
-  # Space Complexity: ?
+  # Time Complexity: O(log n)
+  # Space Complexity: O(1)
   def add(key, value = key)
-    raise NotImplementedError, "Method not implemented yet..."
+    @store << HeapNode.new(key, value)
+    
+    heap_up(@store.length - 1)
   end
-
+  
   # This method removes and returns an element from the heap
   #   maintaining the heap structure
-  # Time Complexity: ?
-  # Space Complexity: ?
+  # Time Complexity: O(log n)
+  # Space Complexity: O(n)
   def remove()
-    raise NotImplementedError, "Method not implemented yet..."
+    return nil if @store.empty?
+    
+    swap(0, -1)
+    result = @store.pop
+    
+    heap_down(0)
+    return result.value
   end
-
-
+  
+  
   # Used for Testing
   def to_s
     return "[]" if @store.empty?
-
+    
     output = "["
     (@store.length - 1).times do |index|
       output += @store[index].value + ", "
     end
-
+    
     output += @store.last.value + "]"
-      
+    
     return output
   end
-
+  
   # This method returns true if the heap is empty
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: O(1)
+  # Space complexity: O(1)
   def empty?
-    raise NotImplementedError, "Method not implemented yet..."
+    return @store.length == 0
   end
-
+  
   private
-
+  
   # This helper method takes an index and
   #  moves it up the heap, if it is less than it's parent node.
   #  It could be **very** helpful for the add method.
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: O(1)
+  # Space complexity: O(1)
   def heap_up(index)
```

Comment:
O(1) for **Both** time and space?  One thing to consider is that you're doing recursion, so that will also incur space complexity and you'll have to heap_up  log n times (max).